internal/core/app_status_report.go

```diff
@@ -0,0 +1,214 @@
+package core
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/zclconf/go-cty/cty"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/go-argmapper"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl/v2"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	sdk "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+	"github.com/hashicorp/waypoint/internal/config"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+// Builds a status report on the given deployment or release
+func (a *App) StatusReport(
+	ctx context.Context,
+	deployTarget *pb.Deployment,
+	releaseTarget *pb.Release,
+) (*pb.StatusReport, *sdk.StatusReport, error) {
+	// Query the artifact
+	var artifact *pb.PushedArtifact
+	if deployTarget != nil && deployTarget.Preload != nil {
+		artifact = deployTarget.Preload.Artifact
+	}
+	if artifact == nil && deployTarget != nil {
+		a.logger.Debug("querying artifact", "artifact_id", deployTarget.ArtifactId)
+		resp, err := a.client.GetPushedArtifact(ctx, &pb.GetPushedArtifactRequest{
+			Ref: &pb.Ref_Operation{
+				Target: &pb.Ref_Operation_Id{
+					Id: deployTarget.ArtifactId,
+				},
+			},
+		})
+		if err != nil {
+			a.logger.Error("error querying artifact",
+				"artifact_id", deployTarget.ArtifactId,
+				"error", err)
+			return nil, nil, err
+		}
+
+		artifact = resp
+	}
+
+	var evalCtx hcl.EvalContext
+	evalCtx.Variables = map[string]cty.Value{}
+	if err := evalCtxTemplateProto(&evalCtx, "artifact", artifact); err != nil {
```

Comment:
If we're targeting a release we actually do need the deployment added here.

See my bigger comment though.